Implementing address to queries

Sam09040 opened the pull request:
## added: 

- address as a field on user type;
- address on user and users query;
- update user's address function on user.db.datasource;
- address seeding;

## updated:

- put graphql folder on api folder;
- put interfaces folder on domain folder;
- put errors, validation and password related files on core folder;
- user and users test to include address related tests;

Sam09040 committed:
Implementing address to queries

- simple resctructuring using core, data, api and domain folders;
- updated user and users tests;
- added function to update a user's address;
- added addresses array to user type;
- added address fetch to user and users query;
- added address seeding;

walteraandrade commented on snaplet/seed/define-seed.ts:
```diff
@@ -24,6 +26,23 @@ export const seedUsers = async (length?: number) => {
         birthDate,
       }),
     );
+    existingAddresses.forEach(async (address) => {
```
isso não está fazendo muito sentido pra mim, se esses campos já existiam e vamos usa-los, pra que apagar o banco de dados e inseri-los de novo? 
além disso, uma função que chama `seedDb` eu não imaginaria que apagaria o banco de ler o nome dela; isso seria um problema de nomenclatura, mas acho que se essa função tem o trabalho de fazer o seed, ela não devia ser responsável por apagar o banco também

Sam09040 replied:
é que, para mim, é importante ter pelo menos um usuário que eu sei as credenciais, ainda mais para fazer o login caso necessário, então, seja em casos que só tem esse usuário ou que tem mais usuários, eu teria como fazer login e realizar outras operações onde eu precisaria de conhecimento das informações de um usuário. Contudo, concordo que ter tudo isso na função de seedDb é delegar muitas responsabilidades para essa função. Vou pensar em um jeito de consertar isso.

walteraandrade commented on src/api/graphql/query/users.query.ts:
```diff
@@ -18,6 +19,11 @@ export const usersQuery = async (skip: number | undefined, limit: number | undef
     limit = 10;
   }
 
+  const addresses = await getAddresses();
```
dentro da users query você adiciona endereços de mock? 🤔 por quê? 🤔

walteraandrade commented on test/users.test.ts:
```diff
@@ -134,6 +145,30 @@ describe('users query', () => {
+    const response = await axios.post(url, { query, variables }, { headers });
+    const data = response.data.data.users.users;
+    expect(data[1]).to.have.property('addresses');
+    expect(data[1].addresses[0]).to.have.property('id');
```
mais valioso que saber se tem esses dados seria saber se eles são aqueles mesmos que esperamos, não?

sophia-bergamo commented on src/api/graphql/query/user.query.ts:
```diff
@@ -0,0 +1,35 @@
+  verifyToken(token);
+
+  const dbUser = await findUserById(id);
+  if (dbUser === null) {
```
```suggestion
  if (!dbUser) {
```

sophia-bergamo commented on src/api/graphql/query/user.query.ts:
```diff
@@ -0,0 +1,35 @@
+    await updateUserAddress(address);
+  }
+
+  const user = await findUserById(id);
```
pq precisa buscar o user aqui de novo? já não temos o `userDb`?

Sam09040 replied:
eu achava que precisaria pegar o usuário de novo depois de atualizar a lista de endereços mas eu percebi que é só atualizar a lista antes.

sophia-bergamo commented on test/users.test.ts:
```diff
@@ -134,6 +145,30 @@ describe('users query', () => {
+    expect(data[1].addresses[0]).to.have.property('state');
+  });
+
   it('should return an error for no users', async () => {
```
pq retornaria um erro se não tiver usuários? acho que a descrição aqui ta errada

Sam09040 replied:
tá mesmo. depois de corrigir que o users retorne uma array vazia caso esteja vazio, esqueci de mudar a descrição.

racsandoval commented on prisma/migrations/20241014114433_user_schema/migration.sql:
```diff
@@ -0,0 +1,13 @@
+-- CreateTable
```
não entendi direito ter essa migration mas ela não ter a tabela de address

Sam09040 replied:
enquanto eu estudava a estrutura do grêmio, eu percebi que vcs deixavam o migrations, então deixei também. o problema é que, como meu notebook deu problema e eu precisei resetar a bios, eu perdi um outro SO que eu tinha no meu computador, então no windows eu só tinha essa migration. eu vou consertar isso.

racsandoval commented on snaplet/seed/define-seed.ts:
```diff
@@ -24,6 +26,23 @@ export const seedUsers = async (length?: number) => {
         birthDate,
       }),
     );
+    existingAddresses.forEach(async (address) => {
```
o `forEach` (e outras funções de array) não respeita ações assíncronas

racsandoval commented on snaplet/seed/define-seed.ts:
```diff
@@ -3,11 +3,13 @@ import { faker } from '@faker-js/faker';
+  const existingAddresses = await getAddresses(existingUser?.id);
 
   const seed = await getSeedClient();
   await resetDatabase(seed);
```
pq é responsabilidade dessa função apagar o banco de dados?

Sam09040 replied:
não é, já que essa função é só para dar seed no banco de dados, então já retirei essa ação dessa função.

racsandoval commented on snaplet/seed/define-seed.ts:
```diff
@@ -24,6 +26,23 @@ export const seedUsers = async (length?: number) => {
     );
+    existingAddresses.forEach(async (address) => {
+      const { id, userId, cep, street, streetNumber, complement, neighborhood, city, state } = address;
+      await seed.address((x) =>
```
o que é `x`?

Sam09040 replied:
o x é uma função do snaplet que ajuda a criar informações similares e é bom para criações múltiplas, mas, como eu tô geralmente criando um usuário/endereço por vez, é melhor eu usar da criação estática.

racsandoval commented on snaplet/seed/define-seed.ts:
```diff
@@ -3,11 +3,13 @@ import { faker } from '@faker-js/faker';
 
-export const seedUsers = async (length?: number) => {
-  const existingUser: User | null = await findUserById(1);
+export const seedDb = async (length?: number) => {
```
dessa forma você não precisa modificar o input nas linhas abaixo (geralmente modificar um input é uma má prática e pode resultar em bugs)
```suggestion
export const seedDb = async (length: number = 50) => {
```

racsandoval commented on snaplet/seed/define-seed.ts:
```diff
@@ -69,6 +117,47 @@ export const seedUsers = async (length?: number) => {
     );
   }
+  let counter = 1;
+  const addresses = Array.from({ length }).map(() => {
```
o segundo argumento do callback da função `map` é o index do elemento da array, não precisaria desse `counter` utilizando ele
```suggestion
  const addresses = Array.from({ length }).map((_, index) => {
```

racsandoval commented on snaplet/seed/define-seed.ts:
```diff
@@ -69,6 +117,47 @@ export const seedUsers = async (length?: number) => {
+  let counter = 1;
+  const addresses = Array.from({ length }).map(() => {
+    counter++;
+    const userId = faker.number.int({ min: 2, max: 51 });
+    const cep = faker.location.zipCode('#####-###');
+    const street = faker.location.street();
+    const streetNumber = faker.location.zipCode('###');
+    const complement = 'Apt. ' + faker.number.int({ min: 100, max: 500 });
+    const neighborhood = faker.location.county();
+    const city = faker.location.city();
+    const state = faker.location.state();
```
pq foram criadas todas essas `consts`? poderia ter adicionado valor direto no objeto, fica mais simples para entender

racsandoval commented on src/api/graphql/query/users.query.ts:
```diff
@@ -18,6 +19,11 @@ export const usersQuery = async (skip: number | undefined, limit: number | undef
   }
 
+  const addresses = await getAddresses();
+  for (const address of addresses) {
+    await updateUserAddress(address);
+  }
```
pq a query de usuários modifica os endereços?

Sam09040 replied:
já removi isso da query de usuários.

Sam09040 committed:
fixes

## removed:
- updateUserAddress from user.db.datasource, user and users query;
- existingUser and existingAddress from seedDb, now using set data;

## updated:
- changed to equal from strings to variables;
- took reset database from seedDb to seed.ts;

racsandoval commented on snaplet/seed/seed.ts:
```diff
@@ -1,2 +1,7 @@
+import { getSeedClient } from './seed-client';
+
+await resetDatabase(await getSeedClient());
+seedDb();
```
o `seedDb` é assíncrono? se sim o log roda antes dessa função terminar

Sam09040 replied:
é assíncrono. já adicionei um await para que o log não rode antes da função terminar.

Sam09040 committed:
asynchronous seedDb

- added await before seedDb on seed.ts so that the log doesn't run before the function ends;
- added await before findUserByEmail on seedDb function;
